packages/grafana-data/src/transformations/transformers/joinDataFrames.ts

```diff
@@ -0,0 +1,230 @@
+  for (let frameIndex = 0; frameIndex < options.frames.length; frameIndex++) {
+    const frame = options.frames[frameIndex];
+    if (!frame || !frame.fields?.length) {
+      continue; // skip the frame
+    }
+
+    let join: Field | undefined = undefined;
+    let fields: Field[] = [];
+    for (let fieldIndex = 0; fieldIndex < frame.fields.length; fieldIndex++) {
+      const field = frame.fields[fieldIndex];
+      getFieldDisplayName(field, frame, options.frames); // cache displayName in state
+
+      if (!join && joinFieldMatcher(field, frame, options.frames)) {
+        join = field;
+      } else {
+        if (options.keep && !options.keep(field, frame, options.frames)) {
+          continue; // skip field
+        }
+
+        let labels = field.labels ?? {};
+        if (frame.name) {
+          labels = { ...labels, name: frame.name };
+        }
+
+        fields.push({
+          ...field,
+          labels, // add the name label from frame
+        });
+      }
+
+      if (options.keepOriginIndices) {
+        field.state!.origin = {
+          frameIndex,
+          fieldIndex,
+        };
+      }
+    }
+
+    if (!join) {
+      continue; // skip the frame
+    }
+
+    if (originalFields.length < 1) {
+      originalFields.push(join); // first join field
+    }
+
+    const a: AlignedData = [join.values.toArray()]; //
+    for (const field of fields) {
+      a.push(field.values.toArray());
+      originalFields.push(field);
+    }
+    allData.push(a);
+  }
+
+  const joined = join(
+    allData,
+    allData.map((v) => v.map((x) => NULL_EXPAND)) // will show gaps in the timeseries panel
+  );
+  if (!joined) {
+    return undefined;
+  }
+
+  return {
+    // ...options.data[0], // keep name, meta?
+    length: joined[0].length,
+    fields: originalFields.map((f, index) => ({
+      ...f,
+      values: new ArrayVector(joined[index]),
+    })),
+  };
+}
+
+//--------------------------------------------------------------------------------
+// Below here is copied from uplot (MIT License)
+// https://github.com/leeoniya/uPlot/blob/master/src/utils.js#L325
+// This avoids needing to import uplot into the data package
+//--------------------------------------------------------------------------------
+
+// Copied from uplot
+type AlignedData = [number[], ...Array<Array<number | null>>];
+
+// nullModes
+const NULL_IGNORE = 0; // all nulls are ignored, converted to undefined (e.g. spanGaps: true)
+const NULL_GAP = 1; // nulls are retained, alignment artifacts = undefined values (default)
+const NULL_EXPAND = 2; // nulls are expanded to include adjacent alignment artifacts (undefined values)
+
+// mark all filler nulls as explicit when adjacent to existing explicit nulls (minesweeper)
+function nullExpand(yVals: Array<number | null>, nullIdxs: number[], alignedLen: number) {
+  for (let i = 0, xi, lastNullIdx = Number.NEGATIVE_INFINITY; i < nullIdxs.length; i++) {
```

Comment:
NEGATIVE_INFINITY was actually a left-over from the previous strategy when i was testing against values. i've changed this to -1 for index testing: https://github.com/leeoniya/uPlot/commit/9324c4263a548a70a1a273e43c07717736e9fd63